source/fuzz/transformation_move_instruction_down.cpp

```diff
@@ -207,13 +249,336 @@ bool TransformationMoveInstructionDown::IsOpcodeSupported(SpvOp opcode) {
+
+      if (ext_inst_import->GetInOperand(0).AsString() != "GLSL.std.450") {
+        return false;
+      }
+
+      switch (static_cast<GLSLstd450>(inst.GetSingleWordInOperand(1))) {
+        case GLSLstd450Round:
+        case GLSLstd450RoundEven:
+        case GLSLstd450Trunc:
+        case GLSLstd450FAbs:
+        case GLSLstd450SAbs:
+        case GLSLstd450FSign:
+        case GLSLstd450SSign:
+        case GLSLstd450Floor:
+        case GLSLstd450Ceil:
+        case GLSLstd450Fract:
+        case GLSLstd450Radians:
+        case GLSLstd450Degrees:
+        case GLSLstd450Sin:
+        case GLSLstd450Cos:
+        case GLSLstd450Tan:
+        case GLSLstd450Asin:
+        case GLSLstd450Acos:
+        case GLSLstd450Atan:
+        case GLSLstd450Sinh:
+        case GLSLstd450Cosh:
+        case GLSLstd450Tanh:
+        case GLSLstd450Asinh:
+        case GLSLstd450Acosh:
+        case GLSLstd450Atanh:
+        case GLSLstd450Atan2:
+        case GLSLstd450Pow:
+        case GLSLstd450Exp:
+        case GLSLstd450Log:
+        case GLSLstd450Exp2:
+        case GLSLstd450Log2:
+        case GLSLstd450Sqrt:
+        case GLSLstd450InverseSqrt:
+        case GLSLstd450Determinant:
+        case GLSLstd450MatrixInverse:
+        case GLSLstd450ModfStruct:
+        case GLSLstd450FMin:
+        case GLSLstd450UMin:
+        case GLSLstd450SMin:
+        case GLSLstd450FMax:
+        case GLSLstd450UMax:
+        case GLSLstd450SMax:
+        case GLSLstd450FClamp:
+        case GLSLstd450UClamp:
+        case GLSLstd450SClamp:
+        case GLSLstd450FMix:
+        case GLSLstd450IMix:
+        case GLSLstd450Step:
+        case GLSLstd450SmoothStep:
+        case GLSLstd450Fma:
+        case GLSLstd450FrexpStruct:
+        case GLSLstd450Ldexp:
+        case GLSLstd450PackSnorm4x8:
+        case GLSLstd450PackUnorm4x8:
+        case GLSLstd450PackSnorm2x16:
+        case GLSLstd450PackUnorm2x16:
+        case GLSLstd450PackHalf2x16:
+        case GLSLstd450PackDouble2x32:
+        case GLSLstd450UnpackSnorm2x16:
+        case GLSLstd450UnpackUnorm2x16:
+        case GLSLstd450UnpackHalf2x16:
+        case GLSLstd450UnpackSnorm4x8:
+        case GLSLstd450UnpackUnorm4x8:
+        case GLSLstd450UnpackDouble2x32:
+        case GLSLstd450Length:
+        case GLSLstd450Distance:
+        case GLSLstd450Cross:
+        case GLSLstd450Normalize:
+        case GLSLstd450FaceForward:
+        case GLSLstd450Reflect:
+        case GLSLstd450Refract:
+        case GLSLstd450FindILsb:
+        case GLSLstd450FindSMsb:
+        case GLSLstd450FindUMsb:
+        case GLSLstd450NMin:
+        case GLSLstd450NMax:
+        case GLSLstd450NClamp:
+          return true;
+        default:
+          return false;
+      }
+    }
+    default:
+      return false;
+  }
+}
+
+bool TransformationMoveInstructionDown::IsMemoryReadInstruction(
+    opt::IRContext* ir_context, const opt::Instruction& inst) {
+  switch (inst.opcode()) {
+    case SpvOpLoad:
```

Comment:
If you think these are relevant feel free to use a TODO(issue) if you don't want to handle them right now.

And the same goes for write instructions I guess: I think there are various instructions that write to images, though I'm not too sure.

Reply:
I'm not very familiar with image-related instructions. What are the most common of them?

Reply:
I am not too familiar with them either, but, from section 3.14 of the spec:

**Instructions that write to images**

- OpImageWrite

**Instructions that read from images:**

- OpImageSampleImplicitLod
- OpImageSampleExplicitLod
- OpImageSampleDrefImplicitLod
- OpImageSampleDrefExplicitLod
- OpImageSampleProjImplicitLod
- OpImageSampleProjExplicitLod
- OpImageSampleProjDrefImplicitLod
- OpImageSampleProjDrefExplicitLod
- OpImageFetch
- OpImageGather
- OpImageDrefGather
- OpImageRead
- OpImageSparseSampleImplicitLod
- OpImageSparseSampleExplicitLod
- OpImageSparseSampleDrefImplicitLod
- OpImageSparseSampleDrefExplicitLod
- OpImageSparseSampleProjImplicitLod
- OpImageSparseSampleProjExplicitLod
- OpImageSparseSampleProjDrefImplicitLod
- OpImageSparseSampleProjDrefExplicitLod
- OpImageSparseFetch
- OpImageSparseGather
- OpImageSparseDrefGather
- OpImageSparseRead

Reply:
Some of these instructions are reserved (e.g. `OpImageSparseSampleProjImplicitLod`). Should I add them too?

Reply:
Should we add support for atomic instructions as well?

Reply:
@Vasniktel Thanks - I see you added the ones that are reserved and I think that's fine.

Would it be easy to add the atomic ones too?  If so then I recommend doing it now, so that we don't forget.

Reply:
Added atomics in b4a76f2.